qiskit/circuit/parameterexpression.py

```diff
@@ -265,14 +265,15 @@ def _apply_operation(
 
         return ParameterExpression(parameter_symbols, expr)
 
-    def gradient(self, param) -> Union["ParameterExpression", float]:
+    def gradient(self, param) -> Union["ParameterExpression", complex, float]:
```

Comment:
Using `complex` alone here as a typehint should suffice. I know things like this have been cleaned up in the past to remove the extra types that are not really needed - its not wrong as such, just not needed. See https://www.python.org/dev/peps/pep-0484/#the-numeric-tower for more info

Reply:
Done in https://github.com/Qiskit/qiskit-terra/pull/6847/commits/a7fd512f9f9011482c55269d83dfe93d4cb38bc1.